src/modules/charging-import/mappers/charge-versions.js

```diff
@@ -0,0 +1,180 @@
+'use strict';
+
+const { isNull, sortBy } = require('lodash');
+const moment = require('moment');
+const dateHelpers = require('../../../lib/date-helpers');
+
+const DATE_FORMAT = 'YYYY-MM-DD';
+
+const STATUS_SUPERSEDED = 'superseded';
+const STATUS_CURRENT = 'current';
+
+const PERIOD_DAY = 'day';
+
+/**
+ * Whether this charge version is considered an "error"
+ * This is either:
+ * - The error flag is set in NALD
+ * - The start date is the same as the following charge version
+ * @param {Object} chargeVersion
+ * @param {Object} nextChargeVersion - next charge version in date/version sequence
+ * @return {Boolean}
+ */
+const isErrorChargeVersion = (chargeVersion, nextChargeVersion) => {
+  const isErrorStatus = chargeVersion.error;
+  const isSameStartDate = nextChargeVersion && (chargeVersion.start_date === nextChargeVersion.start_date);
+  return isErrorStatus || isSameStartDate;
+};
+
+/**
+ * Maps status to equivalent in WRLS
+ * @param {Object} chargeVersion
+ * @param {Number} index - the index in the mapping
+ * @param {Array} arr - the source array in the mapping
+ * @return {Object} charge version updated with mapped status
+ */
+const mapChargeVersionStatus = (chargeVersion, index, arr) => {
+  const nextRow = arr[index + 1];
+
+  return {
+    ...chargeVersion,
+    status: isErrorChargeVersion(chargeVersion, nextRow) ? STATUS_SUPERSEDED : STATUS_CURRENT
+  };
+};
+
+/**
+ * Maps a charge version to a new version where the end date is
+ * modified to prevent it exceeding start date of following CV
+ * @param {Object} chargeVersion
+ * @param {Number} index
+ * @param {Array} source array
+ * @return {Object}
+ */
+const mapChargeVersionEndDate = (chargeVersion, index, arr) => {
+  // If there is no following row, or this row is an error, don't modify date
+  const nextRow = arr[index + 1];
+  if (!nextRow || isErrorChargeVersion(chargeVersion, nextRow)) {
+    return chargeVersion;
+  }
+  const maxEndDate = moment(nextRow.start_date, DATE_FORMAT).subtract(1, PERIOD_DAY).format(DATE_FORMAT);
```

Comment:
I haven't seen it but you might have missed handling charge versions that are invalid in NALD where the start and end dates are the same and the start date is the same as the next charge version start date. If this is the case the line above will set the end date a day before the start date.

Reply:
I have added an additional unit test case for this scenario and it seemed to pass OK.
I also checked the water.charge_versions table from having run locally to check if there are any with the end date before the start, and this seems OK too.